### PR TITLE
Add thread pool and enable threadpool for get features.

### PR DESCRIPTION
**Pull request by wangtt03**
- [X] Forked repo is synced with upstream -> github shows no code delta outside of the desired.
    https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork
- [X] Tests are passing? https://github.com/microsoft/DeepGNN/blob/main/CONTRIBUTING.md#run-tests
- [X] Changelog and documentation updated.
- [X] PR is labeled using the label menu on the right side.

Previous Behavior
----------------
* Get features are using single thread.

New Behavior
----------------
* Add enable threadpool flag in backend and when initializing GE server with thread pool, all the get features requests will use thread pool.

### CODE REVIEW

File: docs/tf/link_pred.rst [hunk 1]
```diff
@@ -120,6 +120,7 @@ Now we can initialize the graph engine in local mode:
     ...    storage_type=PartitionStorageType.memory,
     ...    config_path="",
     ...    stream=False,
+    ...    enable_threadpool=False,
```
Comment (coledie) [inline docs/tf/link_pred.rst:123]: This defaults to false and can be removed right?

File: src/python/deepgnn/tf/common/args.py [hunk 2]
```diff
@@ -34,6 +34,7 @@ def define_param_common(parser):
     group.add_argument("--storage_type", type=lambda type: PartitionStorageType[type], default=PartitionStorageType.memory, choices=list(PartitionStorageType.__members__.keys()) + list(PartitionStorageType), help="Partition storage backing to use, eg memory or disk.")
     group.add_argument("--config_path", type=str, default="", help="Directory where HDFS or other config files are stored.")
     group.add_argument("--stream", action="store_true", default=False, help="If ADL data path, stream directly to memory or download to disk first.")
+    group.add_argument("--enable_threadpool", action="store_true", default=False, help="Whether or not to enable server thread pool.")
```
Comment (coledie) [inline src/python/deepgnn/tf/common/args.py:37]: Can you write somewhere to detail why someone would want to enable the threadpool.

**Review by mortid0 — CHANGES_REQUESTED** [review]
`src/cc/lib/utils/threadpool.h` looks like a copy-paste of https://changkun.de/modern-cpp/exercises/7/7.1/thread_pool.hpp
Please, either write a threadpool from scratch or use a boost threadpool as discussed offline.

File: src/python/deepgnn/graph_engine/snark/client.py [hunk 3]
```diff
@@ -152,6 +152,7 @@ def __init__(
         storage_type: PartitionStorageType = PartitionStorageType.memory,
         config_path: str = "",
         stream: bool = False,
+        enable_threadpool: bool = False,
```
Comment (coledie) [inline src/python/deepgnn/graph_engine/snark/client.py:155]: This is only for server so can we remove it from local ge?

Comment (wangtt03) [reply src/python/deepgnn/graph_engine/snark/client.py:155]: My understanding is that both remote and local GE will use this feature, right? Do we have any concern to use it in local GE?

File: docs/tf/link_pred.rst (hunk 1 above)
Comment (wangtt03) [reply docs/tf/link_pred.rst:123]: We need to add it, otherwise there will be a test break: "enable_threadpool is not in Namespaces".

**Comment by wangtt03** [comment]
> `src/cc/lib/utils/threadpool.h` looks like a copy-paste of https://changkun.de/modern-cpp/exercises/7/7.1/thread_pool.hpp Please, either write a threadpool from scratch or use a boost threadpool as discussed offline.

Thanks @mortid0 . Yes, actually I copied it here: https://github.com/progschj/ThreadPool, but it has some issues in c++17 so I did some updates. Looks like the link you mentioned did the same thing, 😄 anyway, I implemented a new version of thread pool, thanks a lot for point it out.

**Comment by mortid0** [comment]
> Thanks @mortid0 . Yes, actually I copied it here: https://github.com/progschj/ThreadPool, but it has some issues in c++17 so I did some updates. Looks like the link you mentioned did the same thing, 😄 anyway, I implemented a new version of thread pool, thanks a lot for pointing it out.

This implementation is still extremely influenced by that github repo and when you'll clean current code, it will be closer to to that pool. Please either use boost or coroutines instead, otherwise you need to write unit tests for this ThreadPool and handle lots of corner cases.

**Comment by wangtt03** [comment]
> > Thanks @mortid0 . Yes, actually I copied it here: https://github.com/progschj/ThreadPool, but it has some issues in c++17 so I did some updates. Looks like the link you mentioned did the same thing, 😄 anyway, I implemented a new version of thread pool, thanks a lot for pointing it out.
> 
> This implementation is still extremely influenced by that github repo and when you'll clean current code, it will be closer to to that pool. Please either use boost or coroutines instead, otherwise you need to write unit tests for this ThreadPool and handle lots of corner cases.

Updated using boost thread pool, thanks a lot.

File: src/python/deepgnn/graph_engine/snark/client.py (hunk 3 above)
Comment (coledie) [reply src/python/deepgnn/graph_engine/snark/client.py:155]: I see I think argument hint is wrong then, https://github.com/microsoft/DeepGNN/pull/93/files#:~:text=help%3D%22-,Whether%20or%20not%20to%20enable%20server%20thread%20pool,-.%22%2C

Comment (wangtt03) [reply src/python/deepgnn/graph_engine/snark/client.py:155]: Updated, thanks a lot!

File: src/cc/tests/graph_test.cc [hunk 4]
```diff
@@ -358,14 +358,18 @@ TEST_P(StorageTypeGraphTest, NodeFeaturesMultipleNodesSingleFeature)
...
-    std::vector<snark::NodeId> nodes = {0, 1};
-    std::vector<uint8_t> output(4 * 3 * 2);
-    std::vector<snark::FeatureMeta> features = {{0, 12}};
 
-    g.GetNodeFeature(std::span(nodes), std::span(features), std::span(output));
-    std::span res(reinterpret_cast<float *>(output.data()), output.size() / sizeof(float));
-    EXPECT_EQ(std::vector<float>(std::begin(res), std::end(res)), std::vector<float>({1, 2, 3, 5, 6, 7}));
+    for (size_t i = 0; i < 2; i++)
+    {
+        snark::Graph g(path.string(), std::vector<uint32_t>{0}, GetParam(), "", i % 2 == 0);
```
Comment (mortid0) [inline src/cc/tests/graph_test.cc:364]: Please, expand the existing parameterized tests: https://google.github.io/googletest/advanced.html#value-parameterized-tests

File: src/cc/lib/graph/graph.cc [hunk 5]
```diff
@@ -156,22 +188,69 @@ void Graph::GetNodeSparseFeature(std::span<const NodeId> node_ids, std::span<con
...
+    {
+        func(0, node_ids.size(), out_indices, out_data);
+    }
+    else
+    {
+        // data_sections structure:
+        //  - job index
+        //      - feature_index
+        //          - feature_values
+        std::vector<std::vector<std::vector<int64_t>>> indice_sections;
```
Comment (mortid0) [inline src/cc/lib/graph/graph.cc:227]: That's a lot of allocations. Can we avoid them?

**Review by mortid0 — CHANGES_REQUESTED** [review]

File: docs/tf/link_pred.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangtt03) [reply docs/tf/link_pred.rst:122]: We need to add it, otherwise there will be a test break: "enable_threadpool is not in Namespaces".

File: src/cc/lib/benchmark/partition_benchmark.cc [hunk 6]
```diff
@@ -123,13 +124,19 @@ static void BM_NODE_FEATURES_MEMORY(benchmark::State &state)
...
+
 static void BM_NODE_STRING_FEATURES_MEMORY(benchmark::State &state)
 {
     BM_NODE_STRING_FEATURES(state, snark::PartitionStorageType::memory);
 }
 
 BENCHMARK(BM_NODE_FEATURES_DISK)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(1);
 BENCHMARK(BM_NODE_FEATURES_MEMORY)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(1);
+BENCHMARK(BM_NODE_FEATURES_MEMORY_THREADPOOL)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(1);
 BENCHMARK(BM_NODE_FEATURES_DISK)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(2);
```
[the remark this answers is not in the document]
Comment (wangtt03) [reply src/cc/lib/benchmark/partition_benchmark.cc]: Done.

File: src/cc/lib/graph/graph.h [hunk 7]
```diff
@@ -73,12 +74,27 @@ class Graph
...
+        std::function<void(const std::size_t &index, const std::size_t &start_offset, const std::size_t &end_offset)>
+            callback) const;
+
     std::vector<Partition> m_partitions;
     absl::flat_hash_map<NodeId, uint64_t> m_node_map;
     std::vector<uint32_t> m_partitions_indices;
     std::vector<uint64_t> m_internal_indices;
     std::vector<uint32_t> m_counts;
     Metadata m_metadata;
+    std::shared_ptr<boost::asio::thread_pool> m_threadPool;
```
[the remark this answers is not in the document]
Comment (wangtt03) [reply src/cc/lib/graph/graph.h]: Thread pool has been changed to singleton based on the last offline sync up. So, it can be shared by both client and server. Thanks!

File: src/cc/lib/graph/graph.h [hunk 8]
```diff
@@ -24,7 +25,7 @@ class Graph
 {
   public:
     Graph(std::string path, std::vector<uint32_t> partitions, PartitionStorageType storage_type,
-          std::string config_path);
+          std::string config_path, bool enable_threadpool = false);
```
[the remark this answers is not in the document]
Comment (wangtt03) [reply src/cc/lib/graph/graph.h:28]: Thread pool has been changed to singleton and can initialize with thread count now. Thanks!

File: src/python/deepgnn/graph_engine/snark/client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wangtt03) [reply src/python/deepgnn/graph_engine/snark/client.py:155]: I prefer to do it in next PR, this PR focus on `thread pool/dense feature/load partition`, does it make sense?

File: src/cc/tests/graph_test.cc (hunk 4 above)
Comment (wangtt03) [reply src/cc/tests/graph_test.cc]: Thanks a lot, I followed the guide and updated the tests.

File: src/cc/lib/graph/graph.cc (hunk 5 above)
Comment (wangtt03) [reply src/cc/lib/graph/graph.cc]: Removed these extra allocations now, currently we focus on making fixed-size dense features in parallel with thread pool.

File: src/cc/lib/benchmark/partition_benchmark.cc [hunk 9]
```diff
@@ -123,13 +124,19 @@ static void BM_NODE_FEATURES_MEMORY(benchmark::State &state)
...
+}
+
 static void BM_NODE_STRING_FEATURES_MEMORY(benchmark::State &state)
 {
     BM_NODE_STRING_FEATURES(state, snark::PartitionStorageType::memory);
 }
 
 BENCHMARK(BM_NODE_FEATURES_DISK)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(1);
 BENCHMARK(BM_NODE_FEATURES_MEMORY)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(1);
+BENCHMARK(BM_NODE_FEATURES_MEMORY_THREADPOOL)->RangeMultiplier(2)->Range(1 << 3, 1 << 10)->Threads(1);
```
[the remark this answers is not in the document]
Comment (wangtt03) [reply src/cc/lib/benchmark/partition_benchmark.cc]: Sure, here are some benchmark comparisons:

1. Local load partition (node_num/partitions/feature_size):
```
BM_LOAD_GRAPH_NO_THREADPOOL/1024/8/512/iterations:100     4710169 ns      4701002 ns          100
BM_LOAD_GRAPH_THREADPOOL/1024/8/512/iterations:100        4513652 ns      1791682 ns          100

BM_LOAD_GRAPH_NO_THREADPOOL/2048/8/512/iterations:100    10422902 ns     10412854 ns          100
BM_LOAD_GRAPH_THREADPOOL/2048/8/512/iterations:100        8834887 ns      3217835 ns          100

BM_LOAD_GRAPH_NO_THREADPOOL/4096/8/512/iterations:100    22308006 ns     22298785 ns          100
BM_LOAD_GRAPH_THREADPOOL/4096/8/512/iterations:100       17495675 ns      6104018 ns          100

BM_LOAD_GRAPH_NO_THREADPOOL/1024/8/1024/iterations:100    9958105 ns      9938488 ns          100
BM_LOAD_GRAPH_THREADPOOL/1024/8/1024/iterations:100       7534614 ns      1993149 ns          100

BM_LOAD_GRAPH_NO_THREADPOOL/2048/8/1024/iterations:100   18760192 ns     18723498 ns          100
BM_LOAD_GRAPH_THREADPOOL/2048/8/1024/iterations:100      14643855 ns      3538831 ns          100

BM_LOAD_GRAPH_NO_THREADPOOL/4096/8/1024/iterations:100   33711627 ns     33699308 ns          100
BM_LOAD_GRAPH_THREADPOOL/4096/8/1024/iterations:100      28843865 ns      6623976 ns          100
```

2. Local graph (batch_size) (feature size = 602):
```
BM_NODE_FEATURES_MEMORY/32/threads:1                   19230 ns        19230 ns        42056
BM_NODE_FEATURES_MEMORY_THREADPOOL/32/threads:1       284553 ns       252699 ns         2858

BM_NODE_FEATURES_MEMORY/64/threads:1                   55705 ns        55704 ns        10000
BM_NODE_FEATURES_MEMORY_THREADPOOL/64/threads:1       302522 ns       263894 ns         2658

BM_NODE_FEATURES_MEMORY/128/threads:1                 105622 ns       105621 ns         4772
BM_NODE_FEATURES_MEMORY_THREADPOOL/128/threads:1      354288 ns       307630 ns         2414

BM_NODE_FEATURES_MEMORY/256/threads:1                 300937 ns       294931 ns         2833
BM_NODE_FEATURES_MEMORY_THREADPOOL/256/threads:1      406195 ns       340617 ns         2487

BM_NODE_FEATURES_MEMORY/512/threads:1                 524866 ns       524868 ns         1200
BM_NODE_FEATURES_MEMORY_THREADPOOL/512/threads:1      377462 ns       296855 ns         2206

BM_NODE_FEATURES_MEMORY/1024/threads:1               1089031 ns      1089035 ns          607
BM_NODE_FEATURES_MEMORY_THREADPOOL/1024/threads:1     537492 ns       335560 ns         2291
```

3. grpc single server (batch_size/feature_size):
```
BM_DISTRIBUTED_GRAPH_SINGLE_NODE/256/64/iterations:10000                     292458 ns        79063 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/256/64/iterations:10000          382113 ns        62984 ns        10000

BM_DISTRIBUTED_GRAPH_SINGLE_NODE/1024/64/iterations:10000                    675000 ns       107219 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/1024/64/iterations:10000         671018 ns        85958 ns        10000

BM_DISTRIBUTED_GRAPH_SINGLE_NODE/4096/64/iterations:10000                   1881000 ns       136998 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/4096/64/iterations:10000        1856281 ns       117826 ns        10000

BM_DISTRIBUTED_GRAPH_SINGLE_NODE/256/256/iterations:10000                    301045 ns        77823 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/256/256/iterations:10000         474708 ns        77552 ns        10000

BM_DISTRIBUTED_GRAPH_SINGLE_NODE/1024/256/iterations:10000                   721490 ns       103763 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/1024/256/iterations:10000        721390 ns        87149 ns        10000

BM_DISTRIBUTED_GRAPH_SINGLE_NODE/4096/256/iterations:10000                  2328507 ns       148528 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/4096/256/iterations:10000       2058450 ns       121079 ns        10000

BM_DISTRIBUTED_GRAPH_SINGLE_NODE/2048/512/iterations:10000                  1581290 ns       126060 ns        10000
BM_DISTRIBUTED_GRAPH_SINGLE_NODE_THREADPOOL/2048/512/iterations:10000       1302640 ns       101239 ns        10000
```

File: src/cc/lib/distributed/client.cc [hunk 10]
```diff
@@ -418,8 +418,8 @@ void GRPCClient::GetNodeFeature(std::span<const NodeId> node_ids, std::span<Feat
             auto curr_feature_reply = reply.feature_values().c_str();
             for (auto index : reply.offsets())
             {
-                std::copy(curr_feature_reply, curr_feature_reply + fv_size, curr_feature_out + fv_size * index);
-                curr_feature_reply += fv_size;
+                std::copy(curr_feature_reply + fv_size * index, curr_feature_reply + fv_size * (index + 1),
```
Comment (mortid0) [inline src/cc/lib/distributed/client.cc:421]: That looks wrong, `reply.offsets()` can't be always increasing.. Previously we were sequentially copying from `reply.feature_values()`, but you changed it to a random offset..

File: docs/tf/link_pred.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangtt03) [reply docs/tf/link_pred.rst:122]: We just need to do it in this test case (The same as `config_path` and `stream`), because we manually created a "Namespace" object, in actual use case, we don't need to specify it because we already have a default value.

[the remark this answers is not in the document]
Comment (mortid0) [reply docs/tf/link_pred.rst:122]: Another reason to use a hydra for configuration..

File: src/cc/lib/utils/thread_pool.h [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    ThreadPool(ThreadPool &&) = delete;
+    ThreadPool &operator=(ThreadPool &&) = delete;
+
+    // Before calling GetInstance, we can use this to set the desired count of threads
+    // in the thread pool, by default, hardware concurrency is used.
+    static void SetThreadPoolSize(const std::size_t &size);
+    static ThreadPool &GetInstance();
+
+    template <typename _Iterator, typename _Function>
+    void RunInParallel(_Iterator _First, _Iterator _Last, const _Function &_Func)
```
Comment (mortid0) [inline src/cc/lib/utils/thread_pool.h:29]: can we use ranges instead of iterators?

**Comment by coledie** [comment]
Can we close this PR, it can be reopened later? @mortid0 @wangtt03
